scripts/generate.py

```diff
@@ -268,6 +270,48 @@ def generate_moved(destination: str, *, location: str, console: Console):
     with open(destination, "w", newline=newline) as f:
         f.write(rendered_template)
 
+ZIPAPP_MAIN = """\
+#!/usr/bin/env python
+
+import os
+import runpy
+import sys
+
+lib = os.path.join(os.path.dirname(__file__), "lib")
+sys.path.insert(0, lib)
+
+runpy.run_module("pip", run_name="__main__")
```

Comment:
Do we need some sort of minimum python version compatibility check here ? Or is it done by pip somewhere else ?

Reply:
> the main question is whether it's worth it in the first place.

Well, even today we sometimes get support requests for users who manage to run pip with an unsupported python.
With this approach I can see myself wrapping the zipapp in a global `pip` script and launching it from active virtual environments. I'm sure I'll end up myself launching it by accident from older python 2 or python 3.5 environment so...

Reply:
OK, fair point. I'll add it.

Reply:
Done.

Reply:
@sbidoul coming round to this again, the code for the version check here is not compatible with older Python versions. This is the same issue as we had with `__pip-runner__.py`. However, in this case, I'm a lot less comfortable trying to hard code the version as a tuple, because we'd have to either keep it in sync with the pip repository, or do something nasty like try to read the tuple from `__pip-runner__.py`.

What are your thoughts here?

Reply:
I have a version of the check which looks at Python-Requires, and as long as it has the form ">= X.Y" it adds a compatible version check like we do in `__pip-runner__.py`. If the requirement is more complex than that (which hopefully it won't ever be) the code simply omits the check with a warning (in the generate script, not in the zipapp itself).

We'd have to revisit this if we ever needed a more complex Python requirement in pip in the future. Maybe we should therefore abort the generate rather than just warn? But given that we don't generate here until release day, it might be a bit late to find out at that point 🙁

Reply:
Could the zipapp use `__pip_runner__.py` ? (Disclaimer: I've not looked closely at all, just throwing the idea, which maybe complete nonsense)

Reply:
I tried that, and I couldn't work out how. Because `__pip-runner__.py` locates pip from its own location, it has to be run in place. I tried to get `runpy.run_script` to work, but it wouldn't co-operate, and I honestly don't know why :-( And I don't reqlly want to invoke pip via an extra subprocess, that just seems wasteful.

I think it *should* be possible to use `runpy.run_script` for this, but the `runpy` docs are so obscure I find it really hard to work out how to do anything with them :-( I'll do some further tests, though - I might be able to get it to work.

Reply:
No, it appears that neither `runpy.run_script` nor the Python interpreter itself will run a file as `python arc.zip/something.py`. So I think we have to copy the check.

I've updated the check to be 2.x compatible, by parsing a `Requires-Python` of ">=X.Y". I'm happy to improve this, but I don't have any particularly good ideas of what would be better.